xmtp_mls/src/groups/mls_sync.rs

```diff
@@ -351,286 +351,371 @@ where
-                        sender_installation_id = hex::encode(&sender_installation_id),
-                        installation_id = %self.client.installation_id(),group_id = hex::encode(&self.group_id),
-                        current_epoch = mls_group.epoch().as_u64(),
-                        msg_epoch,
-                        msg_group_id,
-                        msg_id,
-                        "[{}] decoding application message",
-                        self.context().inbox_id()
-                    );
-                    let message_bytes = application_message.into_bytes();
-
-                    let mut bytes = Bytes::from(message_bytes.clone());
-                    let envelope = PlaintextEnvelope::decode(&mut bytes)?;
-
-                    match envelope.content {
-                        Some(Content::V1(V1 {
-                                             idempotency_key,
-                                             content,
-                                         })) => {
-                            let message_id =
-                                calculate_message_id(&self.group_id, &content, &idempotency_key);
-                            let queryable_content_fields = Self::extract_queryable_content_fields(&content);
-                            StoredGroupMessage {
-                                id: message_id,
-                                group_id: self.group_id.clone(),
-                                decrypted_message_bytes: content,
-                                sent_at_ns: envelope_timestamp_ns as i64,
-                                kind: GroupMessageKind::Application,
-                                sender_installation_id,
-                                sender_inbox_id,
-                                delivery_status: DeliveryStatus::Published,
-                                content_type: queryable_content_fields.content_type,
-                                version_major: queryable_content_fields.version_major,
-                                version_minor: queryable_content_fields.version_minor,
-                                authority_id: queryable_content_fields.authority_id,
-                                reference_id: queryable_content_fields.reference_id,
-                            }
-                                .store_or_ignore(provider.conn_ref())?
+        Ok(IntentState::Committed)
+    }
+
+    #[tracing::instrument(level = "trace", skip_all)]
+    async fn validate_and_process_external_message(
+        &self,
+        provider: &XmtpOpenMlsProvider,
+        mls_group: &mut OpenMlsGroup,
+        message: PrivateMessageIn,
+        envelope: &GroupMessageV1,
+        cursor: Option<i64>,
+    ) -> Result<(), GroupMessageProcessingError> {
+        let GroupMessageV1 {
+            created_ns: envelope_timestamp_ns,
+            id: ref msg_id,
+            ..
+        } = *envelope;
+
+        let processed_message = mls_group.process_message(provider, message)?;
```

Comment:
If this isn't in a transaction, won't this function immediately delete the keys used to decrypt the message?

What happens if the transaction below rolls back? How do we get those keys back to retry?

Reply:
You're right. I missed this when I was doing read only connection checks. Let me move it into the transaction.

Reply:
This is resolved.